apollo-router/src/services/subgraph_service.rs

```diff
@@ -917,39 +920,69 @@ pub(crate) async fn notify_batch_query(
             // graphql_response, so zip_eq shouldn't panic.
             // Use the tx to send a graphql_response message to each waiter.
             for (response, sender) in rs.into_iter().zip_eq(senders) {
-                sender
-                    .send(Ok(response))
-                    .map_err(|error| FetchError::SubrequestBatchingError {
-                        service: service.to_string(),
-                        reason: format!("tx send failed: {error:?}"),
-                    })?;
+                if let Err(log_error) =
+                    sender
+                        .send(Ok(response))
+                        .map_err(|error| FetchError::SubrequestBatchingError {
+                            service: service.to_string(),
+                            reason: format!("tx send failed: {error:?}"),
+                        })
+                {
+                    tracing::error!(service, error=%log_error, "failed to notify sender that batch processing succeeded");
+                }
             }
         }
     }
 
     Ok(())
 }
 
+type BatchInfo = (
+    (String, http::Request<Body>, Context, usize),
+    Vec<oneshot::Sender<Result<SubgraphResponse, BoxError>>>,
+);
+
 /// Collect all batch requests and process them concurrently
 #[instrument(skip_all)]
 pub(crate) async fn process_batches(
     client_factory: HttpClientServiceFactory,
     svc_map: HashMap<String, Vec<BatchQueryInfo>>,
 ) -> Result<(), BoxError> {
     // We need to strip out the senders so that we can work with them separately.
+    let mut errors = vec![];
     let (info, txs): (Vec<_>, Vec<_>) =
+    // let (info, txs) =
         futures::future::join_all(svc_map.into_iter().map(|(service, requests)| async {
-            let (_op_name, context, request, txs) = assemble_batch(requests).await;
+            let (_op_name, context, request, txs) = assemble_batch(requests).await?;
 
-            ((service, request, context, txs.len()), txs)
+            Ok(((service, request, context, txs.len()), txs))
         }))
         .await
         .into_iter()
+        .filter_map(|x: Result<BatchInfo, BoxError>| x.map_err(|e| errors.push(e)).ok())
         .unzip();
 
+    // If errors isn't empty, then process_batches cannot proceed. Let's log out the errors and
+    // return
+    if !errors.is_empty() {
+        for error in errors {
+            tracing::error!("assembling batch failed: {error}");
+        }
+        return Err(SubgraphBatchingError::ProcessingFailed(
+            "assembling batches failed".to_string(),
+        )
+        .into());
+    }
     // Collect all of the processing logic and run them concurrently, collecting all errors
-    // TODO: Is it ok to panic if the length of the txs and responses does not match? It seems like that is literally impossible
     let cf = &client_factory;
+    // It is not ok to panic if the length of the txs and info do not match. Let's make sure they
+    // do
+    if txs.len() != info.len() {
```

Comment:
I think that this might be literally impossible since we're using `unzip` to construct this, but I'm OK with leaving this if we want to be extra sure.